src/modules/storage/LedgerStorage.ts

```diff
@@ -235,4 +280,381 @@ export class LedgerStorage extends Storages
+     * Put a transaction to database
+     * @param data a transaction data
+     * @param callback If provided, this function will be called when
+     * the database was finished successfully or when an error occurred.
+     * The first argument is an error object.
+     */
+    private putTransaction (block_height: number, tx: any, tx_index: number,
+        tx_hash: string, callback?: (err: Error | null) => void)
+    {
+        if (tx == undefined) {
+            if (callback != undefined)
+                callback(new Error("Parameter validation failed."));
+            return;
+        }
+
+        let sql: string =
+        `INSERT INTO transactions
+            (block_height, tx_index, tx_hash, type, inputs_count, outputs_count)
+        VALUES
+            (?, ?, ?, ?, ?, ?)`;
+        this.db.run(sql,
+            [
+                block_height,
+                tx_index,
+                tx_hash,
+                tx.type,
+                tx.inputs.length,
+                tx.outputs.length
+            ], (err: Error | null) =>
+        {
+            if (err)
+            {
+                if (callback != undefined)
+                    callback(err);
+                return;
+            }
+
+            this.putInputs(block_height, tx, tx_index, tx_hash, (err2: Error | null) =>
+            {
+                if (err2)
+                {
+                    if (callback != undefined)
+                        callback(err2);
+                    return;
+                }
+
+                this.putOutputs(block_height, tx, tx_index, tx_hash, (err3: Error | null) =>
+                {
+                    if (callback != undefined)
+                        callback(err3);
+                });
+            })
+        });
+    }
+
+    /**
+     * Puts all transactions
+     * @param block: block Json data
+     * @param callback If provided, this function will be called when
+     * the database was finished successfully or when an error occurred.
+     * The first argument is an error object.
+     */
+    public putTransactions (block: any, callback?: (err: Error | null) => void)
+    {
+        let idx: number = 0;
+        let doPut = () =>
+        {
+            if (idx >= block.txs.length)
+            {
+                if (callback != undefined)
+                    callback(null);
+                return;
+            }
+
+            this.putTransaction(block.header.height.value, block.txs[idx],
+                idx, block.merkle_tree[idx], (err: Error | null) =>
+                {
+                    if (!err)
+                    {
+                        idx++;
+                        doPut();
+                    }
+                    else
+                    {
+                        if (callback != undefined)
+                            callback(err);
+                        else
+                            return;
+                    }
+                }
+            );
+        }
+        doPut();
```

Comment:
This is a very convoluted pattern I've seen a few times. It was mentioned that it was for asynchronicity ?
I don't quite follow how this helps, there's no `async` or `await` here.

Reply:
I looked for it for a long time because I wanted to handle it like that.
I haven't found a good solution yet.
Since I need to process multiple queries sequentially, I have to get a callback and process them.
I think I can use a function called serial, but we need more time.